File: MATLAB/example_customTWSA.m

% setup some values that we may want to iterate through and save them into
% a structure for convenience.
config.participant = 'Hamner2010_v4_subject01';
config.conditions = {'Run_20002','Run_30002','Run_40002','Run_50002'}
config.trialnums = 1:3

for c = 1%1:4 % you can use parfor if you have the parallel computing toolbox to run all conditions simultaneously
    % uses the Run_20002 condition for the example
    
    %% analyses
    import rraTools.*   
    for t = config.trialnums
        
        mydir = pwd; % get current folder (where code is running from)
        cd('..\') % move up one folder
        % get system path to data folder
        trialpath = fullfile(pwd, 'HamnerOpt\subject01\',config.conditions{c},['Trial_' num2str(t)]);
        cd(mydir) % move back to code folder
        
        r = rrasetup(trialpath,config.participant,[]);
        r.fileset.kinfile = [config.conditions{c}, '_IK.mot'];
        r.fileset.grffile = [config.conditions{c}, '_GRF.mot'];
        r.toolsettings.kinfile = [config.conditions{c}, '_IK.mot'];

        %%      
        r.writeTasksFile('Kp',1600,'UniformWeights',true) % uses defaults, but demonstrates how to specify as inputs
        % write external loads file
        r.writeExtLoads()
        % write ReservesFile to match specified model. Change the value
        % after 'ResidualForce' to increase or decrease the optimal force
        % of residual actuators (FX, FY, FZ, MX, MY, MZ). Change the value
        % after 'ReserveForce' to increase or decrease the ideal torque
        % actuators applied at each joint.
        r.writeReservesFile('ReserveForce',2000,'ResidualForce',75)
        %% Run the first iteration of RRA 
        r = r.runInitialRRA()
        %% Run remaining mass iterations
        r = r.runMassItrsRRA()
        %% setup the TWSA with non-default settings
        
        % get peak net ground reaction force and use to normalize residuals
        % in the optimizer
        maxGRF = r.readPeakExtForce()
        
        % specify kinematic error tolerances for normalization. Between 2
        % and 5 degrees (rotation) and 2-3 cm (translation) is likely
        % appropriate, but may vary between labs and data collection
        % systems.
        rotTol = 5; % specify in degrees. TWSA will convert to rad
        transTol = 0.03; % specify in m.
        
        % run the TWSA with all optional inputs assigned. Overwrite any
        % results currently saved in the folder.
        r.optimizeTrackingWeights('overwrite',true,'min_itrs',10,'max_itrs',200,'fcn_threshold',1,...
            'wRes',3,'wErr',2,'pRes',3,'pErr',2,'ResidualNorm',maxGRF,'RotationNorm',rotTol,'TranslationNorm',transTol)
        % 'overwrite' = true: overwite any TWSA results currently in the folder
        % 'min_itrs' = 10: run at least 10 iterations of RRA
        % 'max_itrs' = 200: don't run more than 200 iterations of RRA
        % 'fcn_threshold' = 1: stop if any iterations produce a function
        % value less than 1 after min_itrs
        %
        % 'wRes' = 3,'wErr' = 2,'pRes' = 3,'pErr' = 2: set cost function
        % weights and polynomials
        %
        % 'ResidualNorm' = maxGRF: normalize residuals based on the peak
        % ground reaction force. 5% of maxGRF for forces, 1% for moments
        %
        % 'RotationNorm' = rotTol: normalize rotational tracking errors by
        % 5 degrees
        %
        % 'TranslationNorm' = transTol: normalize translational tracking
        % errors by 3 cm
    end
end

File: Python/example_customTWSA.py


# %% load the class
import reduceresiduals

# %% setup some variables
# update trial path to match where the test data is saved on your system       
trialpath = "C:/Users/Jordan/Documents/PhD/rra_tools/rra-optimization/HamnerOpt/subject01/Run_40002/Trial_1" # full path to trial folder. must contain scaled model, and ik data
participant = "Hamner2010_v4_subject01" # first part of the model name
condition = [] #empty for this test data, but is otherwise the second part of the model name constructed as "participant_condition.osim"

# %% initialize an instance of the class
rraopts = reduceresiduals.rrasetup( trialpath, participant, condition )


# %% set properties different from defaults
rraopts.fileset.kinfile = "Run_40002_IK.mot"
rraopts.fileset.grffile = "Run_40002_GRF.mot"
rraopts.toolsettings.kinfile = rraopts.fileset.kinfile

# %% specify the reserve actuator (joint torques) and residual actuators optimal force 
rraopts.createReservesFile(ReserveForce=2000, ResidualForce=75)

# %% first pass at RRA. Want to write clean tasks, reserves, and external loads specifications files 
rraopts.initialRRA(createTasks = True, createReserves = False, createExtLoads = True)

# %% perform the mass adjustment iterations
rraopts.runMassItrsRRA()

# %% optional calculate the peak external force (ground reaction force)
# we'll use this value as the basis for normalizing residuals in the TWSA cost function
peak_force = rraopts.readPeakExtForce()


# %% run TWSA with custom settings
rraopts.optimizeTrackingWeights(overwrite = True, min_itrs = 25, max_itrs = 75, fcn_threshold = 2, wRes = 2, wErr = 1, pRes = 3, pErr = 3, ResidualNorm = peak_force, RotationNorm = 5, TranslationNorm = 0.04)
# overwrite = True - deletes saved progress and starts the specified trial from scratch
# min_itrs = 25 - run at least 25 iterations of RRA
# max_itrs = 75 - run at most 75 iterations of RRA 
# fcn_threshold = 2 - terminate TWSA if cost function of any iteration is below this value after at least min_itrs
# wRes = 2, wErr = 1, pRes = 3, pErr = 3 - set cost function weights and polynomials
# ResidualNorm = peak_force - normalize the residuals the cost function using the peak force (5% of ResidualNorm for forces, 1% for moments) 
# RotationNorm = 5 - normalize the rotational tracking errors by 5 degrees in the cost function
# TranslationNorm = 0.04 - normalize the translational tracking errors by 4 cm in the cost function
#


# %%
